.github/workflows/docker.yml

```diff
@@ -1,24 +1,68 @@
 # This workflow will upload a Docker container
 
-name: Upload Docker Container
+name: Build/Upload Docker Container
 
 on:
   workflow_dispatch:
   push:
     branches:
-      - 'master'
+      - '*'
+  pull_request:
 
+env:
+  IMAGE: qmk/base_container
 jobs:
   docker:
     runs-on: ubuntu-latest
     steps:
-    - name: Log in to Docker Hub
-      uses: docker/login-action@v1.10.0
-      with:
-        username: ${{ secrets.DOCKER_HUB_USERNAME }}
-        password: ${{ secrets.DOCKER_HUB_ACCESS_TOKEN }}
-    - name: Build and Push to Docker Hub
-      uses: docker/build-push-action@v2.6.1
-      with:
-        push: true
-        tags: qmkfm/base_container:latest
+      - name: checkout
+        uses: actions/checkout@v2
+
+      - name: Cache Docker layers
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ github.ref }}-${{ github.sha }}
+          restore-keys: |
+            ${{ github.ref }}-${{ github.sha }}
+            ${{ github.ref }}
+            refs/head/master
+
+      - name: create tag
+        run: |
+          SHA=${{ github.sha }}
+          TAG=${IMAGE}:$(TZ=UTC-9 date '+%Y%m')-${SHA:0:7}
+          if [ "${{ github.ref }}" == "refs/heads/master" ]; then
+            LATEST=${IMAGE}:latest
+            echo "TAGS=$LATEST" >> $GITHUB_ENV
+          else
+            echo "TAGS=$TAG" >> $GITHUB_ENV
+          fi
+          echo TAG $TAG
+
+      - name: Log in to Docker Hub
+        if: ${{ github.ref == 'refs/heads/master' }}
+        uses: docker/login-action@v1.10.0
+        with:
+          username: ${{ secrets.DOCKER_HUB_USERNAME }}
+          password: ${{ secrets.DOCKER_HUB_ACCESS_TOKEN }}
+
+      - name: Set up QEMU
+        if: ${{ github.ref == 'refs/heads/master' }}
+        uses: docker/setup-qemu-action@v1
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Build and push
+        id: docker_build
+        uses: docker/build-push-action@v2
+        with:
+          context: ./
+          platforms: ${{ (github.ref == 'refs/heads/master' && 'linux/amd64, linux/arm64') || 'linux/amd64' }}
+          push: ${{ github.ref == 'refs/heads/master' }}
+          tags: ${{ env.TAGS }}
```

Comment:
You should document how these tags work in `README.md` so that people will know what to use.

Reply:
In this implementation, it only gives the "latest" tag when we push the image to the dockerhub.
https://github.com/ryoppippi/qmk_base_container/blob/docker_build/.github/workflows/docker.yml#L35-L37

If the implementation is too confusing, I will omit the place where the tag is defined and just give the "latest" tag.
Do you plan to use git tags in this repoigtory? If so, I'll try to assemble an implementation that uses that tag name.
@skullydazed

Reply:
たった今理解しました。
このイメージは理想的には新しいqmk_cliがリリースされるたびにビルドされるのが一番良いのですね？
そのように変更することも可能ですが、やりましょうか？

Reply:
I just understood.
This image would ideally be best built every time a new qmk_cli is released, right?
I can change it that way, shall I?

Reply:
ok I did it! @skullydazed

Reply:
> In this implementation, it only gives the "latest" tag when we push the image to the dockerhub.

Ok, I wasn't sure if we needed a tag for the arch too. If we just need latest that's perfect.

> This image would ideally be best built every time a new qmk_cli is released, right?
I can change it that way, shall I?

No, we want people to start using `qmkfm/qmk_cli` rather than this container directly. The `qmk_cli` container is built on top of this container.

Reply:
oh i didn't know that change.
ok got it.
After you merge this PR, you can rename CONTAINER_NAME variable as  qmkfm/qmk_cli ;)

Reply:
Anyway, I just changed how to tag built container image!
@skullydazed